### PR TITLE
report: group third-party entities

### PR DESCRIPTION
**Pull request by alexnj**
Work in progress. Requires ~#14614~ #14622 ✅  and #14697 to land first. 

Part 3, "entity grouping and aggregation" changes of [the design doc](https://docs.google.com/document/d/1g9niYq7DNrO44X9xlP89HUh2tIPRXn0j-PKfZDS_CMU/edit#). This introduces grouping of audit results by entities, with aggregated values where applicable. 

Summary of changes:
- [x] All audits that deal with URLs are grouped by entity.
- [x] Introduces a heading/group row to tables that feature: Entity's name, a category chip (or "1st Party") indicating the category of 3P, and a link to open the homepage of 3P (on hover). 
- [x] Report-side aggregate calculations. Columns with units `bytes`, `numeric`, `ms`, `timespanMs` are totaled by default. Can be avoided by `dontAggregate: boolean` flag on a column.
- [x] Allow pre-grouped audit results (covers `ThirdPartySummary` audit).
- [x] Introduce a `sortBy` directive in `Table` and `Opportunity` that indicates to report how aggregation logic should sort rows.

Pending fixes:
- [x] Update base branch to `main` once #14697 lands.
- [x] Fix issue with double-counting third-party group rows in report filter.
- [x] Fix missing category names.
- [x] i18n of strings.
- [x] Update/add tests.
  - [x] Grouping
  - [x] Category name chips groups
  - [x] 1st party chip in groups
  - [x] Legacy non-grouped report and filtering (ensure it's working)
 - [x] Convert `ThirdPartySummary`.entity from a `LinkValue` to a `string` entity name.

Screenshot of _Unused Javascript_ audit report below:

image

### CODE REVIEW

**Comment by connorjclark** [comment]
Can you address the "159 commits" thing? Maybe you need to rebase with origin/main?

**Comment by alexnj** [comment]
> Maybe you need to squash

Done.

File: report/renderer/report-utils.js
Comment (alexnj) [inline report/renderer/report-utils.js:111]: We could output this directly from the audit as well, at the cost of extra LHR payload.

File: core/audits/byte-efficiency/uses-long-cache-ttl.js
Comment (adamraine) [inline core/audits/byte-efficiency/uses-long-cache-ttl.js:278]: What if we make this `aggregate: false` or `aggregateEntities: false` and then give it a default value of `true`?

File: report/renderer/details-renderer.js
Comment (adamraine) [inline report/renderer/details-renderer.js:389]: Is `entityIndexByName` really necessary here? Couldn't we just do `this._entities.list.find(e => e.name === entityName)`?

File: report/renderer/report-utils.js
Comment (adamraine) [inline report/renderer/report-utils.js:16]: Honestly I would prefer to have this information in the LHR rather than do this. This strikes me as something that is going to be technical debt moving forward. What if a plugin want's to change ordering?

Another question is why is this necessary in the first place? Shouldn't the ordering of items in the audit details serve as the canonical ordering that the renderer should use?

File: core/audits/byte-efficiency/uses-long-cache-ttl.js
Comment (alexnj) [reply core/audits/byte-efficiency/uses-long-cache-ttl.js:278]: That would result in additional LHR payload for every column, right? The default falsy and when omitted works to reduce transport.

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:389]: That's an unnecessary `O(n)` scan operation per row rendered, and hence we have the lookup tables.

File: core/audits/byte-efficiency/uses-long-cache-ttl.js
Comment (adamraine) [reply core/audits/byte-efficiency/uses-long-cache-ttl.js:278]: I'm saying that `aggregate: undefined` should be treated the same as `aggregate: true` in this scenario, though making falsy values work this same can also be valuable.

I guess the name `dontAggregate` bugs me because of the don't/dont business. I could also see `noAggregation` or `preventAggregation` working.

File: report/renderer/details-renderer.js
Comment (adamraine) [reply report/renderer/details-renderer.js:389]: How many entities is any page realistically going to have? Adding an extra data structure to the lhr to reduce time complexity seems like splitting hairs IMO.

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js:16]: I'm on the fence regarding the first point, and could be convinced to move this to LHR. The tradeoff is additional LHR payload that serves no value to consumers, and is mostly static in nature. That said, headings in LHR too are static, and yet we've included them in LHR. So this is likely an unnecessary savings optimization.

The reason why this directive is necessary though, is because the sort order of individual items could be different than the sort order post grouping by entities. For example, two individual resources might not have made it to the top in our flat list before, but when grouped together, it could end up being right on top. I.e., ordering of items do change at the reporting side post grouping, and the report renderer has no way of identifying the original sorting logic.

Comment (adamraine) [reply report/renderer/report-utils.js:16]: > and is mostly static in nature

FWIW I don't think we should consider this static if the sort varies between different audits.

> For example, two individual resources might not have made it to the top in our flat list before, but when grouped together, it could end up being right on top.

Sorry not sure I follow. Are you saying the we need the sort order to determine the ordering of entities, even if we could determine the correct order of individual items?

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:389]: I don't think we have the data yet to count the entities that would be detected for an average case yet. Once we ship entity-classification, we should be able to query. FWIW, a page like theverge.com homepage seems to have anywhere between 50-100 entities detected. We could do the helper function.

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js:16]: > even if we could determine the correct order of individual items

Report side doesn't have this information. The `details.item` array is already sorted and ReportRenderer doesn't have a way to figure out which columns were prioritized for sorting.

To make the example more clear, look at this Unused Javascript audit result, pre-grouping:
image

The same audit result post-grouping, would look like this:
image

The entity that the earlier report called out (Concert) is no longer at the top. However, to determine the correct sort columns and their order, we'll need that information be passed over from audit to ReportRenderer. Hope that makes sense?

Comment (alexnj) [reply report/renderer/report-utils.js:16]: I need to find out why the count regressed (15 vs. 16).

Comment (alexnj) [reply report/renderer/report-utils.js]: Distributed this back to audits.

File: core/audits/audit.js
Comment (adamraine) [inline core/audits/audit.js:130]: The way it is now, it seems like this function will actively sort the table when you call it which isn't the case. Could we rename this to `sortedBy`.

File: core/audits/bootup-time.js
Comment (adamraine) [inline core/audits/bootup-time.js:131]: Seems like we could remove these touples and just use the string here. Once we know which value the table is sorted by, it should be easy to determine if it's in ascending or descending order right?

Comment (alexnj) [reply core/audits/bootup-time.js:131]: So far every sort seems to follow the pattern — if it's numeric, it's descending order and if it's alphanumeric, it's ascending. May be this is a fair assumption to make for future as well?

File: report/renderer/details-renderer.js
Comment (adamraine) [inline report/renderer/details-renderer.js:388]: Did we ever resolve the discussion on this?

I still think that if we need to map entities by name, the map should be constructed in the report renderer and not shipped as a part of the LHR.

File: report/renderer/details-renderer.js
Comment (adamraine) [inline report/renderer/details-renderer.js:457]:
```suggestion
    /** @type {string[] */
    const aggregateKeys = [];
```

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:388]: This is a fresh angle and implementation you're bringing up :). We're relying on the entity classification format we shipped earlier. 

FWIW, I don't see an advantage to removing lookup tables and reconstructing it during reporting. I'm assuming you're suggesting to pack origins as a property of each entity, but that makes the structure unusable without post processing (for us and other JSON consumers). What's the downside you see in having lookup tables?

File: types/lhr/audit-details.d.ts
Comment (alexnj) [reply types/lhr/audit-details.d.ts:110]: My bad. I kept shrinking the type through refactors and at this point it should've been removed. 😄

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:388]: The difference is wasted computation cycles each time we have to access or do something with the structure. Neither structures are friendly to human consumption of this data, so I've traded to optimize on performance and converting O(n) lookups to O(1) via lookup tables, while minimizing redundancy of information. Sure the perf in terms of a single LightHouse manual run is trivial, but it could make a difference in repeated runs if anyone's doing the report at scale in the cloud.

**Comment by alexnj** [comment]
Move sum/aggregate values to detail level as:

skipSumming: string[] | undefined
skipGrouping: boolean

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:388]: Okay, so I'll start making those changes as a separate PR and rebase this against that.

File: report/renderer/details-renderer.js
Comment (paulirish) [inline report/renderer/details-renderer.js:436]:
```suggestion
     // In subitem-situations (unused-js), ensure Entity name is not rendered as code, etc.
    entityColumnHeading.valueType = 'text';
```

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:388]: https://github.com/GoogleChrome/lighthouse/pull/14744

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:537]: This seems correct, we want row + subrows to be the same shade.

File: report/renderer/report-utils.js
Comment (connorjclark) [inline report/renderer/report-utils.js:151]: This has nothing to do with table items. It's pretty generic. Maybe `getComparatorForKeys`/`getComparatorForProperties`?

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:450]:
```suggestion
      if (supportedAggregations.includes(heading.valueType)) {
```

File: report/renderer/details-renderer.js
Comment (adamraine) [inline report/renderer/details-renderer.js:512]: Could this be done in `_renderEntityGroupRow`?

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:454]: early return if `aggregateKeys` is empty?

File: report/renderer/details-renderer.js
Comment (adamraine) [inline report/renderer/details-renderer.js:468]: Not sure if this would ever happen, but I think it makes sense to be defensive here. If `item[key]` is undefined for some reason this is `NaN`.

```suggestion
        aggregate[key] = Number(aggregate[key] || 0) + Number(item[key] || 0);
```

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:437]: I'm having difficulty grasping this function. Could you add some tests for this explicitly? is this right:

"Returns an array of aggregated table items to use as the top-level rows in an aggregated table. Each table item returned represents a unique entity, with every applicable key that can be aggregated as a property. Optionally, these table items are sorted by specified keys."

**Review by adamraine — COMMENTED** [review]
Noticed a bug:


https://user-images.githubusercontent.com/6752989/216413587-a46b0799-4a8a-403c-b071-0a2385fc8b37.mov

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:457]: I'm not clear on the function of primary key, can you elaborate? Does this enforce the first heading of an audit to meet a specific constraint to be grouped like this?

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:517]: Good thought to use the existing zebra-striping on `.lh-row--group`, makes the table aggregation clearer. 👍 

It's unfortunate that we'll have a different UI approach for separating items based on if aggregated or not, though (when not aggregated: zebra. otherwise: nothing...). Items with long subitems next to each other really benefit from this striping, which is lost here. so I think it's worth exploring this more. What are some other options? think border-bottom? some sort of iconography before the entity name?

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:385]: nit: `chip` as a term only comes up in this function. Possible to drop extra terminology? just `_adornEntityTableRow`?

File: report/renderer/details-renderer.js
Comment (connorjclark) [inline report/renderer/details-renderer.js:499]: It'd be helpful to come up with a name for the table item we use for the entity descriptor row, and use it throughout `_renderEntityGroupRow` / `_adornTableRowWithEntityChips` / this function

**Review by connorjclark — COMMENTED** [review]
Flushing first pass!

File: types/lhr/audit-details.d.ts
Comment (alexnj) [reply types/lhr/audit-details.d.ts:80]: Yeah you're right. That was the previous iteration; we decided to move all such parameters to `detail` level which keeps all of them at one place.

File: report/renderer/details-renderer.js
Comment (connorjclark) [reply report/renderer/details-renderer.js:457]: Could things become simpler if the Table/Opp details had a `aggregationHeading?: {keys: string[], sortBy?: string[], displayUnit?: string, granularity?: number}`? This would match the existing `subItemsHeading`.

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:512]: Yep.

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:517]: Yeah we tried a few combinations of these with Peter/UX, and you can see the other attempts [here](https://docs.google.com/presentation/d/1iXm0xCe_Bd5lAipIGYdv5lWup1H6OpICa254-ZKQiCs/edit#slide=id.g18cc1102f3a_2_32), in the slides around it. This is cosmetic, so we could iterate if we have better options.

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js:151]: So one thought I have is — we might want to show "1st Party" group as the topmost group, regardless of its values, to acknowledge the actionability aspect. That would make it very specific to a comparator for the table/group rows. Thoughts?

File: types/lhr/audit-details.d.ts
Comment (connorjclark) [reply types/lhr/audit-details.d.ts:112]: Ah, so it is representing how the audit does any sorting, like byte-efficiency will be sorted by wastedBytes, so the rows witihin an aggregate entity should be sorted the same way...


isn't this the same as sorting by `detail.items.indexOf(item)`?

File: report/renderer/details-renderer.js
Comment (adamraine) [reply report/renderer/details-renderer.js:457]: `aggregationHeading?: {keys: string[], sortBy?: string[]}` sounds good unless we plan to use the `sortBy` information elsewhere besides 3p stuff. I was imagining a future where `sortBy` passed to `makeTableDetails` will actively sort the items without manually doing it elsewhere in the audit. In that case it would make sense to keep them separate.

That was just an idea though, there are clearly multiple ways to do this. I'll defer to @alexnj on this one.

Comment (connorjclark) [reply report/renderer/details-renderer.js:457]: > [re: hypothetical use case] without manually doing it elsewhere in the audit

It's gotta be declared in the audit anyway. Either with a `sortBy` property, or a call to something like `Audit.makeTableDetail(..., {sortBy: []})`. I don't see why the former would be preferred over the latter.

File: report/renderer/details-renderer.js
Comment (connorjclark) [reply report/renderer/details-renderer.js:517]: thanks, I left a comment there.

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:457]: I expect more of the aggregation to move to audit time, over time. I like the idea of `makeTableDetails` doing the sorting (at least for the simple cases, which is the majority). 

What would `keys: string[]` contain? The format isn't really going match the existing heading/subHeading structure, so I see it as just moving from one structure to another. `aggregationHeading: false` doesnt necessary translate to `isEntityGrouped`, one indicates a directive to group items, and the latter indicates that the audit result already comes grouped.

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:500]: One thing to note is that all render* prefixed functions return a DOM/DocumentFragment. I think it might be better off calling this `_getEntityGroup...Items`

File: report/renderer/details-renderer.js
Comment (alexnj) [reply report/renderer/details-renderer.js:459]: No we need to continue grouping, even if we aren't summing any columns.

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js]: Apparently we need to do that since we're doing arithmetic, and TypeScript isn't happy about it.

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js:155]: What would be another type we'd want to sort as a column value?

Comment (connorjclark) [reply report/renderer/report-utils.js:155]: Want to? Probably no others. It costs nothing to support bools here. The others are a bit annoying. Mostly, I think there should be an explicit warning somewhere (even if just a console log in the report) if trying to sort by an unsupported column type. That could be login here, in the report. Or validation on the audit runner side (where we throw an error more safely, given we would quickly catch that in development)

File: report/renderer/report-ui-features.js
Comment (adamraine) [inline report/renderer/report-ui-features.js:251]: nit: just semantics at this point but could we change this to check for `lh-row--even`/`lh-row-odd` instead. The current logic assumes the only alternative to grouping is zebra styling which could theoretically change in the future.

File: report/renderer/report-ui-features.js
Comment (adamraine) [inline report/renderer/report-ui-features.js:268]: Can you add a test case that ensures the zebra striping doesn't sneak in when we use the 3p checkbox

**Review by adamraine — APPROVED** [review]
Overall I think LGTM!

File: report/renderer/report-utils.js
Comment (alexnj) [reply report/renderer/report-utils.js:155]: Added a console warning + tests. Please have a look.

File: report/test/renderer/details-renderer-test.js
Comment (adamraine) [inline report/test/renderer/details-renderer-test.js:634]: I think you'll want to do this in an `after(() => ...)` clause in case `renderer.render` throws an error

**Review by connorjclark — CHANGES_REQUESTED** [review]
I think the PR looks pretty good, my main concern is related to the design of the table, esp re: zebra striping as I lined out in my other comment. I also think there's some risk involved with landing such a large change just before cutting a major release, and want us all to have at least a few weeks of hands on experience with this before shipping to users. 10.1 seems like a reasonable target.

**Comment by adamraine** [comment]
There are changes to the lhr like `sortBy`, `skipSumming` etc. I think it would be good to land those changes in 10.0 so that the 10.1 report renderer can handle 10.0 lhrs.

**Comment by connorjclark** [comment]
Those are additions, not breaking changes. 10.1 renderer (assuming with this PR) would just render 10.0 LHR as normal without aggregation, no? Doesn't seem critical to get those in for 10.0.

We expedite 10.1 a bit (~3 weeks instead of 4-6) and still make the same Chromium cut, I believe. So to avoid CDT having a chrome release of 10.0 instead of 10.1. Is that your concern?

File: core/audits/audit.js
Comment (connorjclark) [reply core/audits/audit.js:132]: This seems worth working out first.

If we consider `Audit.makeTableDetails` part of our stable API (we probably should), then to defer this change post-10.0 we will need to support both all the params one-by-one, plus a single options object (or maybe `headings, results, opts` for just the optionals, but at that point should probably just collapse all of them into the object). Checking at runtime the number of parameters (or whatever) wouldn't be a big deal, and actually doing a breaking change here will just needlessly break custom audits, so I can see us doing that indefinitely.

Whatever approach we do, we certainly don't want to release a version that has 2 params + 4 optional params, making us have to then support 3 interfaces here indefinitely. So I'd consider this important to address before landing this PR (though, the work to make this a options object should certainly be its own PR)

Comment (alexnj) [reply core/audits/audit.js:132]: What should be the signature of the refactored function `makeTableDetails(heading, results, options)`, with summary collapsed into the object? Who is depending on this other than LH/audit code?

**Comment by adamraine** [comment]
Yeah I guess it's not breaking. SGTM

File: core/audits/audit.js
Comment (connorjclark) [reply core/audits/audit.js:132]: Custom audits.

> makeTableDetails(heading, results, options) ?

That, or just one options param for everything. Up for debate.

Comment (alexnj) [reply core/audits/audit.js:132]: https://github.com/GoogleChrome/lighthouse/pull/14753

Once finalized, I'll adjust this PR accordingly - should be trivial.

**Comment by alexnj** [comment]
I've updated this branch with [latest UX recommendations](https://docs.google.com/document/d/1_Eqk1iQLzsor-7TEQaaXjyKXsomYgXzbhl4cxv82Ee4/edit?resourcekey=0-rn_Vd2c3J9AOKn9IF4xaRA#) so it should be good for review/merge.

**Review by connorjclark — APPROVED** [review]

**Pull request merged**
